Add new fix method to set variable with element value #282

TobiasNx opened the pull request:
Resolves #282

blackwinter reviewed (changes requested):
I've updated the implementation to account for missing values and added tests for edge cases (22e0482).

Open questions:

1. How to handle missing values? Is an empty string the right choice?
2. How to handle non-string values? Is the Java object representation the right choice? (Previously, you [said](https://github.com/metafacture/metafacture-fix/issues/282#issuecomment-2355583039) that it "should only work with string-elements".)

Finally, the new function needs to be documented.

TobiasNx commented:
> I've updated the implementation to account for missing values and added tests for edge cases ([a52dc9b](https://github.com/metafacture/metafacture-fix/commit/a52dc9bd3c897d7c6461ba6846e3b2819faa653b)).
> 
> Open questions:
> 
>     1. How to handle missing values? Is an empty string the right choice?
I would not add a variable if the element does not exist.

> 
>     2. How to handle non-string values? Is the Java object representation the right choice? (Previously, you [said](https://github.com/metafacture/metafacture-fix/issues/282#issuecomment-2355583039) that it "should only work with string-elements".)
I thought MF only processes string values in the Stream-Processing. Because of that we introduced marker for numbers in the json-encoder. The idea with string was the connection to the other ticket in mf-core but I have put the idea in core on hold.

> 
> 
> Finally, the new function needs to be documented.

blackwinter commented:
> I thought MF only processes string values in the Stream-Processing.

Metafix knows data types strings, arrays and objects; variables can only contain strings.

TobiasNx commented:
> > I thought MF only processes string values in the Stream-Processing.
> 
> Metafix knows data types strings, arrays and objects; variables can only contain strings.

Ah, I forgot. Then I would suggest an exception that the function expected a string but got hash or array.

blackwinter commented:
> I would not add a variable if the element does not exist.

This might become difficult to use:

1. If one record has the field and the next one doesn't, the variable keeps its previous value. We might want to remove the variable instead.
2. Referencing a non-existing variable (if we remove it or if it just hasn't been set yet) results in an error. So you can't reliably use the variable. We might have to introduce a conditional that explicitly checks for the presence of a variable.

TobiasNx commented:
> > I would not add a variable if the element does not exist.
> 
> This might become difficult to use:
> 
>     1. If one record has the field and the next one doesn't, the variable keeps its previous value. We might want to remove the variable instead.
> 
>     2. Referencing a non-existing variable (if we remove it or if it just hasn't been set yet) results in an error. So you can't reliably use the variable. We might have to introduce a conditional that explicitly checks for the presence of a variable; or add an option to specify the default value.

I favor the the option to define a default value. The conditional for the a variable sounds good. I am also in favor to delete the variable at the end of each record.

blackwinter commented:
> Then I would suggest an exception that the function expected a string but got hash or array.

Implemented in cee638a.

> I favor the the option to define a default value.

Implemented in 46d1ef5.

> The conditional for the a variable sounds good. I am also in favor to delete the variable at the end of each record.

Should probably be handled in separate issues.

TobiasNx commented:
@TobiasNx I added documentation. Should @fsteeg do a code review?

blackwinter commented on README.md:
```diff
@@ -267,6 +267,23 @@ put_vars(
+
+Options:
+
+- `"default"`: provide default value if source field does not exist.
```
Drop quotes, capitalize first word after colon.

Option description should say what it _is_, not what it _does_.

```suggestion
- `default`: Default value if source field does not exist.
```

TobiasNx replied:
Isn't it important to highlight that this option has to be in quotes?

blackwinter replied:
See [discussion in #354](https://github.com/metafacture/metafacture-fix/pull/356#discussion_r1799506405).

TobiasNx replied:
```suggestion
- `default`: Default value if source field does not exist. The option needs to be written in quotation marks because it is a reserved word in Java. (Default: Empty string)
```

blackwinter commented on README.md:
```diff
@@ -267,6 +267,23 @@ put_vars(
+
+- `"default"`: provide default value if source field does not exist.
+
+
```
Extra empty line.

TobiasNx replied:
```suggestion
```

TobiasNx commented on README.md:
```diff
@@ -267,6 +267,23 @@ put_vars(
 
+##### `to_var`
+
+Defines a single global variable that can be referenced with `$[<variableName>]` and assinges the value of the `<sourceField>`.
```
```suggestion
Defines a single global variable that can be referenced with `$[<variableName>]` and assigns the value of the `<sourceField>`.
```

TobiasNx replied:
Done. Thanks.

TobiasNx commented on README.md:
````diff
@@ -267,6 +267,23 @@ put_vars(
+Defines a single global variable that can be referenced with `$[<variableName>]` and assinges the value of the `<sourceField>`.
+
+```perl
+to_var("<sourceField>", "<variableValue>")
````
```suggestion
to_var("<sourceField>", "<variableName>")
```

TobiasNx replied:
Done. Thanks.

fsteeg reviewed (approved):
Very useful feature!
